core/src/commonIntegrationTest/kotlin/com/powersync/DatabaseTest.kt

```diff
@@ -114,6 +230,71 @@ class DatabaseTest {
             }
         }
 
+    @Test
+    fun testClosingReadPool() =
+        runTest {
+            val pausedLock = CompletableDeferred<Unit>()
+            val inLock = CompletableDeferred<Unit>()
+            // Request a lock
+            val lockJob =
+                async {
+                    database.readLock {
+                        inLock.complete(Unit)
+                        runBlocking {
+                            pausedLock.await()
+                        }
+                    }
+                }
+
+            // Wait for the lock to be active
+            inLock.await()
+
+            // Close the database. This should close the read pool
+            // The pool should wait for jobs to complete before closing
+            val closeJob =
+                async {
+                    database.close()
+                }
+
+            // Wait a little for testing
+            // Spawns in a different context for the delay to actually take affect
+            async { withContext(Dispatchers.Default) { delay(500) } }.await()
+
+            // The database should not close yet
+            assertEquals(actual = database.closed, expected = false)
```

Comment:
Can we assert that, in this state, calling `database.readLock` again throws?

Reply:
Come to think of it, it's weird that `database.closed` is false but we can't use the database because it's closing. I'm not sure if there should be a `isClosing` getter in addition to `closed` or if it doesn't matter.

Reply:
I've added a check to assert that new calls to readLock fail while the database is closing. 
The concept of an isClosing state is interesting. I think we can add that if there are any issues in practice. For now some read lock requests might acquire a lock during the initial phase of closing. In which case the close operation will wait for that lock operation to complete.